xds/internal/balancer/clusterresolver/e2e_test/balancer_test.go

```diff
@@ -302,3 +313,227 @@ func (s) TestErrorFromParentLB_ResourceNotFound(t *testing.T) {
+		lbCfgCh:      make(chan serviceconfig.LoadBalancingConfig, 1),
+		Builder:      b,
+		ConfigParser: b.(balancer.ConfigParser),
+	}
+}
+
+func (b *wrappedPriorityBuilder) Build(cc balancer.ClientConn, opts balancer.BuildOptions) balancer.Balancer {
+	priorityLB := b.Builder.Build(cc, opts)
+	return &wrappedPriorityBalancer{
+		Balancer:  priorityLB,
+		scStateCh: b.scStateCh,
+		lbCfgCh:   b.lbCfgCh,
+	}
+}
+
+type wrappedPriorityBalancer struct {
+	balancer.Balancer
+	scStateCh *buffer.Unbounded
+	lbCfgCh   chan serviceconfig.LoadBalancingConfig
+}
+
+func (b *wrappedPriorityBalancer) UpdateSubConnState(sc balancer.SubConn, state balancer.SubConnState) {
+	b.scStateCh.Put(state)
+	b.Balancer.UpdateSubConnState(sc, state)
+}
+
+func (b *wrappedPriorityBalancer) UpdateClientConnState(ccs balancer.ClientConnState) error {
+	select {
+	case b.lbCfgCh <- ccs.BalancerConfig:
+	default:
+	}
+	return b.Balancer.UpdateClientConnState(ccs)
+}
+
+func (b *wrappedPriorityBalancer) Close() {
+	b.scStateCh.Close()
+	b.Balancer.Close()
+}
+
+// Test verifies that SubConn state changes are propagated to the child policy
+// by the cluster resolver LB policy.
+func (s) TestSubConnStateChangePropagationToChildPolicy(t *testing.T) {
+	// Unregister the priority balancer builder for the duration of this test,
+	// and register a policy under the same name that makes SubConn state
+	// changes pushed to it available to the test.
+	priorityBuilder := balancer.Get(priority.Name)
+	internal.BalancerUnregister(priorityBuilder.Name())
+	testChildPolicy := newWrappedPriorityBuilder(priorityBuilder)
+	balancer.Register(testChildPolicy)
+	defer balancer.Register(priorityBuilder)
+
+	managementServer, nodeID, bootstrapContents, _, cleanup := e2e.SetupManagementServer(t, e2e.ManagementServerOptions{})
+	defer cleanup()
+
+	server := stubserver.StartTestService(t, nil)
+	defer server.Stop()
+
+	// Configure cluster and endpoints resources in the management server.
+	resources := e2e.UpdateOptions{
+		NodeID:         nodeID,
+		Clusters:       []*v3clusterpb.Cluster{e2e.DefaultCluster(clusterName, edsServiceName, e2e.SecurityLevelNone)},
+		Endpoints:      []*v3endpointpb.ClusterLoadAssignment{e2e.DefaultEndpoint(edsServiceName, "localhost", []uint32{testutils.ParsePort(t, server.Address)})},
+		SkipValidation: true,
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	if err := managementServer.Update(ctx, resources); err != nil {
+		t.Fatal(err)
+	}
+
+	// Create xDS client, configure cds_experimental LB policy with a manual
+	// resolver, and dial the test backends.
+	cc, cleanup := setupAndDial(t, bootstrapContents)
+	defer cleanup()
+
+	// Spawn a goroutine which ensures that SubConn state changes are propagated
+	// to the child policy. We *only* wait for READY state since we expect to
+	// get to that state before the below RPC succeeds.
+	doneCh := make(chan error, 1)
+	go func() {
```

Comment:
Curious, now that it's a buffer.Unbounded and the write doesn't block, you could technically do this after line 434 in main test goroutine right? The sc updates get pushed to buffer in a non blocking way then you read after?

Reply:
Good point. Done.